binary_search_tree/tree.py

```diff
@@ -8,52 +8,132 @@ def __init__(self, key, val = None):
     def add(self, key, value = None):
-        pass
+        new_node = TreeNode(key, value)
+        if self.root is None:
+            self.root = new_node
+            return new_node
+        
+        self.add_helper_func(self.root, new_node)
+    
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(log n) if tree is balanced, O(n) if unbalanced
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        if self.root is None:
+            return None
 
-    # Time Complexity: 
-    # Space Complexity: 
+        current = self.root
+
+        while current is not None:
+            if key < current.key:
+                current = current.left
+            elif key > current.key:
+                current = current.right
+            else:
+                return current.value
+        return None
+
+    def inorder_helper_func(self, current_node, items):
+        if current_node is not None:
+            self.inorder_helper_func(current_node.left, items)
+            items.append({"key": current_node.key, "value": current_node.value})
+            self.inorder_helper_func(current_node.right, items)
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        items = []
+        self.inorder_helper_func(self.root, items)
+        return items
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def preorder_helper_func(self, current_node, items):
+        if current_node is not None:
+            items.append({"key": current_node.key, "value": current_node.value})
+            self.preorder_helper_func(current_node.left, items)
+            self.preorder_helper_func(current_node.right, items)
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)    
     def preorder(self):
-        pass
+        items = []
+        self.preorder_helper_func(self.root, items)
+        
+        return items
 
-    # Time Complexity: 
-    # Space Complexity:     
+
+    def postorder_helper_func(self, current_node, items):
+        if current_node is not None:
+            self.postorder_helper_func(current_node.left, items)
+            self.postorder_helper_func(current_node.right, items)
+            items.append({"key": current_node.key, "value": current_node.value})
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
     def postorder(self):
-        pass
+        items = []
+        self.postorder_helper_func(self.root, items)
+        
+        return items
+
+    def height_helper_func(self, current_node, height):
+        if current_node is not None:
+            left= self.height_helper_func(current_node.left, height)
+            right= self.height_helper_func(current_node.right, height)
+            return max(left, right) + 1
+        
+        return 0
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)     
     def height(self):
-        pass
```

Comment:
✨ However, space complexity is actually O(log n) if the tree is balanced and O(n) if unbalanced because of the recursive stack call. Similarly, time complexity would be O(log n) for a balanced tree